server/src/main/java/org/elasticsearch/cluster/NodeConnectionsService.java

```diff
@@ -99,7 +100,9 @@ public void connectToNodes(DiscoveryNodes discoveryNodes, Runnable onCompletion)
         final List<Runnable> runnables = new ArrayList<>(discoveryNodes.getSize());
         try (var refs = new RefCountingRunnable(onCompletion)) {
             synchronized (mutex) {
-                for (final DiscoveryNode discoveryNode : discoveryNodes) {
+                // Ugly hack: when https://github.com/elastic/elasticsearch/issues/94946 is fixed, just iterate over discoveryNodes here
+                for (final Iterator<DiscoveryNode> iterator = discoveryNodes.mastersFirstStream().iterator(); iterator.hasNext();) {
```

Comment:
If we are going to use this pattern in multiple places, what do you think about creating a helper method that yields an iterable object that represents master first sorted discoveryNodes?

```java
public Iterable<DiscoveryNode> mastersFirst() {
        return mastersFirstStream()::iterator;
}
```

Then we can continue using for loops as 
```java
for (final DiscoveryNode discoveryNode : discoveryNodes.mastersFirst()) {
```

Reply:
Thanks, yes, that'd make sense. I think it's just here for now so I'll leave it as it is to avoid adding too much more cruft to the `DiscoveryNodes` API, but I will do this if I encounter any more of these.